### PR TITLE
Add jupyter-rfb recipe

### PR DESCRIPTION
**Pull request by djhoese**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by djhoese** [comment]
@almarklein Please comment to confirm that you are willing to be a maintainer for this conda-forge package.

**Comment by djhoese** [comment]
Looks like the osx job failed because of an HTTP hiccup. I will wait for the other jobs to see if I need to make more changes otherwise I will close and reopen this to restart the CI jobs.

@almar do you think there are any simple tests that we should add to the testing section to make sure this is built properly?

**Comment by djhoese** [comment]
I'm really not sure how to fix the build. I thought adding jupyterlab would help, but now I get this on OSX:

```
Could not build wheels for jupyter-rfb which use PEP 517 and cannot be installed directly
```

My understanding is that this is probably from an old version of pip. It could also be some dependencies missing on either pyproject.toml or the setup.py.

**Comment by almarklein** [comment]
> @almarklein Please comment to confirm that you are willing to be a maintainer for this conda-forge package.

Sure!

> @almarklein do you think there are any simple tests that we should add to the testing section to make sure this is built properly?

The biggest risk of "not built properly" might be that it does not include the JS. But at the moment we don't have any tests that take the JS into account. On the other hand. Our setup.py was made with the cookie cutter, so ...

**Comment by djhoese** [comment]
> The biggest risk of "not built properly" might be that it does not include the JS. But at the moment we don't have any tests that take the JS into account. On the other hand. Our setup.py was made with the cookie cutter, so ...

Does the sdist include the necessary javascript files? I think it is weird that this is working on windows and linux but not mac. Maybe missing npm or yarn or something.

**Comment by almarklein** [comment]
Cant we make this is a noarch build?

**Comment by almarklein** [comment]
Oh. It is a noarch build.

**Comment by almarklein** [comment]
> Does the sdist include the necessary javascript files?

Yes.

**Comment by almarklein** [comment]
Anything [in here](https://github.com/conda-forge/ipyvolume-feedstock/blob/master/recipe/meta.yaml) that we can learn from?

**Comment by djhoese** [comment]
Ha that's actually what I've been looking at and why I added jupyterlab. One thing that is interesting is that this is a `noarch: python` package but is being built on all three platforms. The ipyvolume feedstock only gets built for linux (since it is noarch).

**Comment by djhoese** [comment]
Only other thing I could try is the jupyter-packaging version specifier/limit.

**Comment by djhoese** [comment]
I apparently broke CI.

**Comment by djhoese** [comment]
@conda-forge-admin, please restart ci

**Comment by djhoese** [comment]
See https://github.com/conda-forge/conda-forge-ci-setup-feedstock/pull/159 which was pointed to on gitter. I assume restarting CI is enough.

**Comment by djhoese** [comment]
@conda-forge/help-python This is ready for review.

File: recipes/jupyter-rfb/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+build:
+  number: 0
+  noarch: python
+  script: {{ PYTHON }} -m pip install . -vv
+
+requirements:
+  host:
+    - jupyter-packaging >=0.7.9,<0.8
+    - jupyterlab >=3.0,<4
```
Comment (bollwyvl) [inline recipes/jupyter-rfb/meta.yaml:21]: unless you really need it, it's often most robust to _not_ require `jupyterlab` at build time. Luckily `jupyter-packaging` is no longer a runtime dep of `lab`, but... things happen.

also, if you _only_ support `jupyterlab >=X`, but don't want to make it an explicit dependency (which is fine), you can specify:

```yaml
requirements:
  run_constrained:
    - jupyterlab >=X
```

File: recipes/jupyter-rfb/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  run:
+    - ipywidgets >=7.6.0
+    - numpy
+    - python >=3.6
+
+test:
+  imports:
+    - jupyter_rfb
+  commands:
+    - pip check
```
Comment (bollwyvl) [inline recipes/jupyter-rfb/meta.yaml:33]: meanwhile, it probably _is_ worth depending on `jupyterlab` for the test phase, and doing some "smoke" tests with the [labextension listing](https://github.com/search?q=org%3Aconda-forge+labextension+list&type=code) mechanims to verify your assets get deployed to `$PREFIX/share` and/or `$PREFIX/etc` and the metadata (e.g. `install.json`) is right.

File: recipes/jupyter-rfb/meta.yaml (hunk 1 above)
Comment (djhoese) [reply recipes/jupyter-rfb/meta.yaml:21]: Thanks. I didn't know about run_constrained. I'll try that out. I think I only added the jupyterlab dependency after the builds weren't working the first time.

File: recipes/jupyter-rfb/meta.yaml (hunk 2 above)
Comment (djhoese) [reply recipes/jupyter-rfb/meta.yaml:33]: Yes, of course. Good point. I'll copy over some of those examples.

**Comment by djhoese** [comment]
Ok I've updated the tests and the deps. I based it off of the dask-labextension but we'll see how this works as far as `_` versus `-`. On my system all the files produced by the extension (including the `jupyter labextension list` output) use `_`. The dask recipe has a couple hyphens. We'll see.

**Comment by djhoese** [comment]
@bollwyvl Thanks for reviewing. I see you updated the dask-labextension feedstock for the checks of plugin.json. Any idea if that file should always exist? I can't seem to get it to be produced on my local system even though the extension works fine in jupyter lab.

**Comment by bollwyvl** [comment]
If you don't have any users settings (which seems plausible, since it's a widget), `schemaDir` in `package.json` and `schema/plugin.json`, you won't need to check for those... if you _do_ have them, you _really_ need to make sure it's there, as the client isn't very robust to not having that schema file available.

**Comment by djhoese** [comment]
Ah ok. Nope this extension is relatively "stupid" in that sense, just a simple widget passing data to the browser and passing events from the browser. Most recent commit should fix all of this.

**Comment by djhoese** [comment]
:facepalm: And now I'm getting `pip._internal.exceptions.InstallationError: Could not build wheels for jupyter-rfb which use PEP 517 and cannot be installed directly` on Windows (was getting it on OSX before). I also see `  Exception: Jupyter command `jupyter-labextension` not found.` further up in the log. @bollwyvl is this a problem in the conda-forge build or do I actually need to include jupyterlab in my host/build deps?

**Comment by djhoese** [comment]
@bollwyvl The Windows build seems to fail without jupyterlab as a dependency as it wants to run the `jupyter-labextension` command and says it can't find it (see my last comment). Any other ideas for how to get that to work without putting the jupyterlab requirement in the `host` section?

**Comment by bollwyvl** [comment]
Welp, looking through other labextensions (from that search link above) can give you some ideas.

The culprit is likely `jupyter-packaging`... on various extensions, which require various versions of that, it's been necessary to:
- pass `--install-option='--skip-npm'` to `pip install` (this is what lab, itself, does)
- set an environment variable? sometimes people customize them
- delete the `pyproject.toml` as it doesn't really help much for `conda-build`ing, since we're _already in an isolated environment_
- patch the `setup.py` to _just git rid_ of all the special cmdclass stuff

On extensions where I'm not beholden to "toeing the line," and using All Jupyter Stuff, I just dispense with `jupyter-packaging` entirely, as I _don't_ value hiding a peril-fraught npm/webpack build inside a `pip` wheel build... and _really_ don't like doing it inside `conda-build` (for precisely the reasons you have independently discovered and verified) and _reallllly_ don't like doing `nodejs` stuff on Windows.

**Comment by djhoese** [comment]
So I'm a little conflicted on this. This package, jupyter-rfb, has no other purpose than to be a jupyter extension. So getting rid of jupyter-specific stuff seems odd to me. The structure of the project is based off of the jupyter-widget cookiecutter: https://github.com/jupyter-widgets/widget-cookiecutter. If anything should be changed in our project to make it easier to build here then it should probably be brought up on that repository.

I assume the `--skip-npm` means that none of the javascript stuff is done at all, but it also looks like the cookiecutter includes all of those files in the sdist so we shouldn't need them, right? Let me try some things...

**Comment by djhoese** [comment]
@bollwyvl You win. I copied dask-labextension and move to separate build scripts that delete the pyproject.toml and then do `--skip-npm`. Seems like using `pyproject.toml` doesn't pass along `--install-option` flags so this seemed like the only way.

**Comment by bollwyvl** [comment]
Great artists steal!

As has been mentioned... somewhere on one of those extensions: this is more of a `conda-build` problem, as PEP 517 isn't going anywhere, and use of `pyproject.toml` (vs the present, widespread mishmash of it, `setup.cfg`, _and_ `setup.py` even on a single project).

There is also some concern about the continued viability of `data_files` (on which the jupyter stack's ability to interoperate with other language runtimes is basically _built_).

In short: on `conda-forge`, we do what we must, because we can! :cake:

**Comment by djhoese** [comment]
Yeah, doesn't seem perfect, but it passes.

@bollwyvl is there anything else left for this PR/recipe?

**Comment by bollwyvl** [comment]
@conda-forge/staged-recipes @conda-forge/help-python ready for review
